[Editor] Add listPrimEntities function

dbierek opened the pull request:
This PR adds a function that is used by Glacier2Obj and NavKit to get the transforms for all requested GeomEntities.

dbierek commented:
Going to add support for reading includes and excludes for `[modules:/zpfboxentity.class].pc_entitytype`s

dbierek commented:
Ready for review. If you approve, feel free to squash merge if you want to avoid cluttering the commit history.

OrfeasZ commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+    std::vector<std::string> s_selectorPrimHashes;
+    // Keep iterating through the tree until we find all the prims.
+    while (!s_NodeQueue.empty()) {
+        if (entities.size() >= 10) {
+            s_SendEntitiesCallback(entities, false);
+            entities.clear();
+        }
```
What's the purpose of this check and the clearing here?

dbierek replied:
The purpose is to start sending entities quickly in batches, and reduce memory usage by only storing up to 10 entities at a time.

dbierek replied:
Added clarifying comments.

OrfeasZ commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+                            const auto s_AlocHash = s_ReferenceResourceInfo.rid.GetID();
+                            std::string s_AlocHashString{ std::format("{:016X}", s_AlocHash) };
+                            s_AlocHashes.push_back(s_AlocHashString);
+                            Logger::Info("Found ALOC. ID: {} TBLU: {} ALOC: {}", s_Id, s_HashString, s_AlocHashString);
```
This (and some other logs) probably need to be debug logs.

OrfeasZ commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+                    Quat s_EntityQuat = GetQuatFromProperty(s_Node->Entity);
+                    Quat s_ParentQuat = GetParentQuat(s_Node->Entity);
+
+                    Quat s_CombinedQuat;
+                    s_CombinedQuat = s_ParentQuat * s_EntityQuat;
```
This assignment could probably be moved to the variable declaration.

dbierek replied:
Combined declaration and assignment.

OrfeasZ commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+
+                    Quat s_CombinedQuat;
+                    s_CombinedQuat = s_ParentQuat * s_EntityQuat;
+                    std::tuple<std::vector<std::string>, Quat, ZEntityRef> s_Entity =
+                        std::make_tuple(
+                            s_AlocHashes,
+                            s_CombinedQuat,
+                            s_Node->Entity
+                        );
```
This and some of the variables above could use `auto`, since their type can be inferred from the right hand side when reading through the code.

dbierek replied:
Changed throughout to use auto when the right hand side makes it clear what the type should be.

OrfeasZ commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+        char* s_EntityType = s_Interfaces[0].m_pTypeId->typeInfo()->m_pTypeName;
+
+        if (strcmp(s_EntityType, s_PFSEEDPOINT_TYPE) == 0) {
+            Quat s_EntityQuat = GetQuatFromProperty(s_Node->Entity);
+            Quat s_ParentQuat = GetParentQuat(s_Node->Entity);
+
+            Quat s_CombinedQuat;
+            s_CombinedQuat = s_ParentQuat * s_EntityQuat;
+            std::tuple<std::vector<std::string>, Quat, ZEntityRef> s_Entity =
+                std::make_tuple(
+                    std::vector<std::string>{ "00280B8C4462FAC8" },
+                    s_CombinedQuat,
+                    s_Node->Entity);
```
Same notes as above here.

dbierek replied:
Switched to auto.

OrfeasZ commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+            continue;
+        }
+        char* s_EntityType = typeInfo->m_pTypeName;
+        if (strcmp(s_EntityType, s_GEOMENTITY_TYPE) == 0) {
```
Might be a good idea to move the code inside this if block to a separate function, so it's easier to read. Same for the else if below.

dbierek replied:
Will work on extracting these to functions.

dbierek replied:
Extracted these functions.

OrfeasZ commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+        char* s_EntityType = s_Interfaces[0].m_pTypeId->typeInfo()->m_pTypeName;
+
+        if (strcmp(s_EntityType, s_PFBOXENTITY_TYPE) == 0) {
+            Quat s_EntityQuat = GetQuatFromProperty(s_Node->Entity);
+            Quat s_ParentQuat = GetParentQuat(s_Node->Entity);
+
+            Quat s_CombinedQuat;
+            s_CombinedQuat = s_ParentQuat * s_EntityQuat;
+            std::tuple<std::vector<std::string>, Quat, ZEntityRef> s_Entity =
+                std::make_tuple(
+                    std::vector<std::string>{ "00724CDE424AFE76" },
+                    s_CombinedQuat,
+                    s_Node->Entity);
```
Same notes as above here.

dbierek replied:
Made auto.

OrfeasZ commented on Mods/Editor/Src/Editor.h:
```diff
@@ -52,7 +52,12 @@ class Editor : public IPluginInterface {
     std::shared_ptr<EntityTreeNode> GetEntityTree() { return m_CachedEntityTree; }
     void UnlockEntityTree() { m_CachedEntityTreeMutex.unlock_shared(); }
     ZEntityRef FindEntity(EntitySelector p_Selector);
+    std::string getCollisionHash(auto s_SelectedEntity);
```
Different casing than the rest of the functions. Also, `s_` prefix is used for scoped / local variables. This should be prefixed with `p_` for parameters. Similarly below.

OrfeasZ replied:
Also, can this be static?

dbierek replied:
Fixed casing and prefix throughout, and made functions static,

OrfeasZ commented on Mods/Editor/Src/EditorServer.cpp:
```diff
@@ -201,6 +202,34 @@ void EditorServer::OnMessage(WebSocket* p_Socket, std::string_view p_Message) no
         SendEntityList(p_Socket, Plugin()->GetEntityTree(), s_MessageId);
         Plugin()->UnlockEntityTree();
     }
+    else if (s_Type == "listAlocEntities") {
+        Plugin()->LockEntityTree();
+        Plugin()->FindAlocs([p_Socket](std::vector < std::tuple< std::vector<std::string>, Quat, ZEntityRef >> s_Entities, bool s_Done) -> void {
+            SendEntitiesDetails(p_Socket, s_Entities, s_Done);
+        });
+        Plugin()->UnlockEntityTree();
+    }
+    else if (s_Type == "listPfBoxEntities") {
+        Plugin()->LockEntityTree();
+        std::vector<std::tuple<std::vector<std::string>, Quat, ZEntityRef>> s_Entities = Plugin()->FindPfBoxEntities();
+        SendEntitiesDetails(p_Socket, s_Entities, true);
+        Plugin()->UnlockEntityTree();
+    }
+    else if (s_Type == "listPfSeedPointEntities") {
+        Plugin()->LockEntityTree();
+        std::vector<std::tuple<std::vector<std::string>, Quat, ZEntityRef>> s_Entities = Plugin()->FindPfSeedPointEntities();
+        SendEntitiesDetails(p_Socket, s_Entities, true);
+        Plugin()->UnlockEntityTree();
+    }
```
Do these need to be separate commands? I haven't looked at the navkit side, but if you're always calling these together, it might make sense to combine them into a single command because a lot of their logic (e.g. iterating through the tree and checking for specific entities) is duplicated. 

Also, you acquire a lock both here and inside the `FindX` functions. Not sure if that's necessary. It could also be a good idea to automatically populate the entity tree if you detect it hasn't already been populated.

dbierek replied:
* Now rebuilds the tree if it's not built.
* Combined the FindPfBoxEntities and FindPfSeedPointEntities functions. FindAlocs is a little too different since it has some batching logic and handles multiple types of entities.
* Switched to a single socket message that calls the find functions.

OrfeasZ commented on Mods/Editor/Src/EditorServer.cpp:
```diff
@@ -703,12 +732,223 @@ void EditorServer::SendEntityDetails(WebSocket* p_Socket, ZEntityRef p_Entity, s
+    if (!p_Entities.empty()) {
+        Logger::Info("Sending entities details for: '{}' entities", p_Entities.size());
+
+        for (std::tuple<std::vector<std::string>, Quat, ZEntityRef> p_Entity: p_Entities) {
```
This could probably be rewritten as `for (auto& [s_AlocHashes, s_Quat, s_Entity] : p_Entities)`

dbierek replied:
Rewrote to use templated variables.

OrfeasZ commented on Mods/Editor/Src/Editor.h:
```diff
@@ -123,6 +128,10 @@ class Editor : public IPluginInterface {
 
     void SMatrix43Property(const std::string& p_Id, ZEntityRef p_Entity, ZEntityProperty* p_Property, void* p_Data);
 
+    auto* GetProperty(ZEntityRef p_Entity, ZEntityProperty* p_Property);
+    Quat GetQuatFromProperty(ZEntityRef p_Entity);
+    Quat GetParentQuat(ZEntityRef p_Entity);
```
Can these be static too?

dbierek commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+            const uint16_t s_TypeAlignment = s_PropertyInfo->m_pType->typeInfo()->m_nTypeAlignment;
+
+            const std::string s_TypeName = s_PropertyInfo->m_pType->typeInfo()->m_pTypeName;
+            const std::string s_InputId = std::format("##Property{}", i);
```
Removed unused variables.

dbierek commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+                    if (std::string(s_PropertyNameView.Data, s_PropertyNameView
+                        .Size) != s_COLLISION_RESOURCE_ID_PROPERTY_NAME) {
+                        continue;
+                    } else {
```
Removed else to reduce nesting.

dbierek commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+    std::vector<Quat> s_ParentQuats;
+    while (s_Entity->m_eidParent != NULL) {
+        s_EidParent = s_Entity->m_eidParent;
+        std::string s_Id = std::format("{:016x}", s_EidParent.m_ref->GetType()->m_nEntityId);
```
Removed unused.

dbierek commented on Mods/Editor/Src/Components/EditorAPI.cpp:
```diff
@@ -88,6 +89,376 @@ ZEntityRef Editor::FindEntity(EntitySelector p_Selector) {
+    const uint16_t s_TypeAlignment = s_PropertyInfo->m_pType->typeInfo()->m_nTypeAlignment;
+
+    // Get the value of the property.
+    auto* s_Data = (*Globals::MemoryManager)->m_pNormalAllocator->AllocateAligned(s_TypeSize, s_TypeAlignment);
```
Will work on fixing this.

dbierek replied:
Switched this to use unique_ptr with a custom templated deleter.

dbierek commented on Mods/Editor/Src/EditorServer.cpp:
```diff
@@ -703,12 +732,223 @@ void EditorServer::SendEntityDetails(WebSocket* p_Socket, ZEntityRef p_Entity, s
+    p_Socket->send("Done loading Navp.", uWS::OpCode::TEXT);
+}
+
+int GetPropertyValue(const auto* s_Property, ZEntityRef& p_Entity) {
```
Renamed this function to `IsPropertyValueTrue`.

dbierek commented:
@OrfeasZ  I addressed all the PR comments and playtested it with my latest changes to NavKit on the Editor-mod-support branch: https://github.com/glacier-modding/NavKit/tree/Editor-mod-support

There is one persistent issue though, if you load into a mission, export to NavKit, then leave the mission, the next time you try to start a mission Hitman will crash. This has been a problem for months now, but I'm not sure how to solve it. I was hoping fixing that memory leak you identified would fix it, but it seems like there's something else going on.
